docs: Rewrite guidelines for writing commit messages.

alya opened the pull request:
The focus is on making the instructions more specific and organized, so that they are easier to follow.

Current version: https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages

## Decisions for feedback

- Dropped this point, because it seems like it would be too easy to end up with meaningless descriptions like, "I clicked around to test."
   - any manual testing you did in addition to running the automated tests
- Dropped this point, because it seemed to me like it belongs more in the PR description; I added a paragraph about PR descriptions vs. commit messages:
  - any aspects of the commit that you think are questionable and
    you'd like special attention applied to.
- Dropped this point, because it seemed specific to a very small percentage of contributor PRs:
  - If the commit makes performance improvements, you should generally
  include some rough benchmarks showing that it actually improves the
  performance.
- Picked https://github.com/zulip/zulip/commit/084dd216f017c32e15c1b13469bcbc928cd0bce9 as an example commit, but happy to have that replaced with something else

## Follow-ups

- Someone could pick out additional examples of good commit messages. I think this is best done by an engineer, and shouldn't be a blocker for merging the changes.

alya commented:
@laurynmm Would you like to review this one?

alya commented:
Also, @amanagr , @sahil839 , tagging you for feedback, since you review a lot of contributor PRs. Are there any common mistakes we aren't addressing here?

timabbott commented:
I added an extra commit that we should plan to squash with both some wording tweaks/revisions for the earlier sections, as well as some extended examples and a couple new subsections on what belongs in a commit description.

This surely needs some copy-editing.

alya commented on docs/contributing/commit-discipline.md:
```diff
@@ -115,26 +115,36 @@ future contributors, and is no less important than the code you write. This
+Common examples include: settings, message feed, compose, left
+sidebar, right sidebar, recent (for **Recent conversations**), search,
+markdown, tooltips, popovers, drafts, integrations, email, docs, help,
+and api docs; issue trackers can be a good source of inspiration (but
```
What does this mean? Are you referring to labels in GitHub?

alya replied:
(this bit about issue trackers should be either written more clearly or dropped)

laurynmm replied:
I think this was a reference to, if a PR fixes/addresses part of an issue in GitHub, then the issue title could be used as inspiration.

alya replied:
dropping for now

alya commented:
Thanks! I took a copyediting pass.

gnprice commented on docs/contributing/commit-discipline.md:
```diff
@@ -86,87 +86,224 @@ check out GitHub's excellent [blog post](https://github.blog/2022-06-30-write-be
+Many decisions may be documented in multiple places (for example, both
+in a commit message and a code comment). The general rules of thumb are:
+
+- If the information will be helpful long-term for someone trying to
+  read and understand the code base, include it in a code comment.
+- If the information will be helpful long-term for someone trying to
+  investigate why these changes are a complete and correct
+  implementation of the commit's idea, include it in the commit
+  message.
```
This part I feel could be a bit clearer on the distinction between these two concepts. Here's a version adapted from what I have [in the zulip-mobile style guide](https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#commit-messages-code):
```suggestion
- Use the commit message for information that's relevant for someone
  trying to understand the change this commit is making, or the difference
  between the old version of the code and the new version. In particular,
  this includes information about why the new version of the code is better than,
  or not worse than, the old version.
- Use code comments, or the code itself, for information that's relevant
  for someone trying to read and understand the new version of the code
  in the future, without comparing it to the old version.
```

alya replied:
That looks great to me.

timabbott replied:
Yeah that's great!

gnprice reviewed (commented):
This is great, thank you!

laurynmm commented on docs/contributing/commit-discipline.md:
```diff
@@ -86,87 +86,224 @@ check out GitHub's excellent [blog post](https://github.blog/2022-06-30-write-be
+  discussed above) _every_ commit is expected to update tests and documentation
+  as needed.
+- Make it readable to someone who is familiar with Zulip's code base, but hasn't
+  been involved with the project you're working on.
```
We use "project" to refer to Zulip as a whole a few times before this, so maybe here specify a bit more ...

> ... but hasn't been involved with the specific area of the code base you're working on.

or 

> ... but hasn't been involved with the specific area of the project you're working on.

alya replied:
Using "effort", since maybe that's closer to what @timabbott had in mind?

laurynmm commented on docs/contributing/commit-discipline.md:
```diff
@@ -86,87 +86,224 @@ check out GitHub's excellent [blog post](https://github.blog/2022-06-30-write-be
+  variable or function is a better path to readable code than writing
+  a prose explanation.
+- If the information describes an additional change that you made while working
+  on the commit, consider whether it is separable from the rest of the project.
```
Again, I feel like the use of "project" is confusing here. It refers to the changes in the commit vs Zulip as a whole.

alya replied:
replaced with "changes"

laurynmm commented on docs/contributing/commit-discipline.md:
```diff
@@ -86,87 +86,224 @@ check out GitHub's excellent [blog post](https://github.blog/2022-06-30-write-be
+  If it is, it should probably be moved to its own commit, with its own commit
+  message explaining it. It is far easier to review and integrate a series of
+  several well-written commits than it is to review those same changes in a
+  single commit.
```
Felt like a lot of "it" in this bullet point, so revising the last sentence to be ...

> Reviewing and integrating a series of several well-written commits is far easier than reviewing those same changes in a single commit.

laurynmm reviewed (commented):
This looks great overall! I had a few minor points for clarifying what we're talking about in a couple of instances. I like Greg's clarification text above as well.

alya commented:
I've addressed the feedback from @gnprice and @laurynmm above, so I think this is ready for integration review. All the changes I've made since @timabbott 's changes are in one `[squash]` commit.

timabbott commented:
Squashed and merged, thanks @alya for doing this and everyone else who helped review and provide feedback!
